### PR TITLE
(#2297) Apply console context to deprecate warnings

### PR DESCRIPTION
**Pull request by cmaher**
To ensure console compatibility

Had to split out console so I could stub it and test it.

### CODE REVIEW

**Comment by paulfalgout** [comment]
@cmaher jscs error in the spec.

Otherwise :+1:

File: src/helpers.js [hunk 1]
```diff
@@ -145,5 +145,9 @@ var deprecate = Marionette.deprecate = function(message, test) {
   }
 };
 
-deprecate._warn = typeof console !== 'undefined' && (console.warn || console.log) || function() {};
+deprecate._console = typeof console !== 'undefined' ? console : {};
+deprecate._console.warn = deprecate._console.warn || deprecate._console.log || function() {};
```
Comment (jridgewell) [inline src/helpers.js]: You're overwriting `console.warn` here. Not horrible, but it doesn't seem like the best solution.

Comment (cmaher) [inline src/helpers.js]: Only setting it if it's not a thing... but it's still not great to write to a global.

Comment (jasonLaster) [inline src/helpers.js]: would something like this do the trick?

```js
deprecate._warn = function() {
  var fnc = (console && console.warn || console.warn) || function() {};
  fnc.apply(console || window, arguments);
}
```

Comment (cmaher) [inline src/helpers.js]: that's what I'm thinking of, but still using `deprecate._console`.  The indirection is really just there for testing purposes

**Comment by cmaher** [comment]
something seems wrong with travis and jscs

**Comment by jasonLaster** [comment]
indentation lol Expected indentation of 8 characters at test/unit/collection-view.spec.js :

File: src/helpers.js (hunk 1 above)
Comment (jridgewell) [inline src/helpers.js]: What's wrong with just binding to `console`? I don't think it's really necessary to test that `deprecate._warn` calls `console.warn`, only that `deprecate._warn` is called.


```javascript
deprecate._warn = typeof console !== 'undefined' ? _.bind(console.warn || console.log, console) : function() {};
```

Comment (cmaher) [inline src/helpers.js]: Because `deprecate._warn` is called on `deprecate`, and you can't test that `console` is bound.

Comment (jridgewell) [inline src/helpers.js]: But this is a line that should never change? I see the benefit of mocking as 1% and the cost to support as 99%.

Comment (cmaher) [inline src/helpers.js]: Yep, there's just that bad feeling about a bugfix without a test.

**Comment by cmaher** [comment]
updated to not modify console object and for jscs

File: test/unit/collection-view.spec.js [hunk 2]
```diff
@@ -310,7 +311,8 @@ describe('collection view', function() {
           foo: 'biz'
         }, {
           foo: 'baz'
-      }]);
+        }
+      ]);
```
Comment (jasonLaster) [inline test/unit/collection-view.spec.js]: Fucking jscs

File: test/unit/helpers.spec.js [hunk 3]
```diff
@@ -61,10 +61,26 @@ describe('normalizeUIKeys', function() {
 
 describe('Marionette.deprecate', function() {
   beforeEach(function() {
-    this.sinon.stub(Marionette.deprecate, '_warn');
+    this.sinon.spy(Marionette.deprecate, '_warn');
+    this.sinon.stub(Marionette.deprecate, '_console', {
+      warn: this.sinon.stub()
+    });
```
Comment (jasonLaster) [inline test/unit/helpers.spec.js]: Nice dude!

**Comment by jasonLaster** [comment]
:thumbsup:

**Comment by jasonLaster** [comment]
@marionettejs/marionette-core needs another thumb

**Comment by ahumphreys87** [comment]
:+1:

**Comment by ahumphreys87** [comment]
manually merged

**Comment by megawac** [comment]
Shame this didn't make it into 2.4.3 or 2.4.4

**Comment by samccone** [comment]
it can make it into the next release right?

**Comment by paulfalgout** [comment]
It's in for v3.  I assume this was merged in the days where we thought v3 was sooner than it has been.  Currently there is no `2.4.5`

**Comment by megawac** [comment]
I don't understand why we wouldn't do very simple maintenance (such as
bumping backbone versions) on old major versions. No need to force users to
upgrade so aggressively

**Comment by paulfalgout** [comment]
I don't disagree.

**Comment by paulfalgout** [comment]
:cherries: 'd on to `patch`

**Pull request closed without merging**